### ISSUE TITLE
bug: plugin gets loaded multiple times

### ISSUE THREAD

**Issue by greenpau** [epoch8/airflow-exporter#58]
I added logging to the module. Then, I realized that the plugin is being loaded on every task.

The issue is that exporter.py should include only classes and functions. Then, in init.py, we use the classes and functions. But ... it will happen only once. Currently, it is not implemented like that.

**Comment by elephantum** [comment]
@greenpau can you describe in more details pls? what do you mean by "loaded on every task"?

If it means that plugin is loaded on run of each task, then it is an acceptable behaviour, because by default Airflow spawns new airflow process for each task run, so full cycle of initialisation is done.

**Comment by greenpau** [comment]
> If it means that plugin is loaded on run of each task, then it is an acceptable behaviour, because by default Airflow spawns new airflow process for each task run, so full cycle of initialisation is done.

@elephantum, it is not an intended behavior. I just tested by moving the code to `__init__.py`. When you will review the PR #57  you will see that the class will be loaded only once.

**Comment by greenpau** [comment]
@elephantum , you are right that it will load a new process. I am trying to catch that and detect if it is a spawned process or it is main process. If it is spawned process, then I will not load the exporter.

**Comment by elephantum** [comment]
@greenpau do you really need to do that? there's no harm in loading exported that is never called. It does not use significant resources.

**Comment by greenpau** [comment]
> there's no harm in loading exporter that is never called.

@elephantum, If I need to load exporter configuration, I would prefer to do it once.
In current master, the exporter does not have a configuration file. There is no harm there. However, when I added a configuration parameters, it matters.

**Comment by greenpau** [comment]
@elephantum , here is what I came up with to load it only for airflow webserver processes.

https://github.com/epoch8/airflow-exporter/blob/39cbac19fb63ffe5c2627bd5907bbc120154269f/airflow_exporter/__init__.py#L1

https://github.com/epoch8/airflow-exporter/blob/39cbac19fb63ffe5c2627bd5907bbc120154269f/airflow_exporter/__init__.py#L242-L244

**Comment by elephantum** [comment]
I don't like reliance on `argv` contents to determine if the plugin should be loaded, this seems extremely fragile to me.

As we don't have any heavy logic in plugin init, I see no harm in loading it multiple times.

Closing with "wont fix".

**Comment by twbecker** [comment]
Does this explain why the plugin appears multiple times in the list in the UI and also why the "Metrics" item in the Admin menu appears multiple times?

**Comment by elephantum** [comment]
@twbecker no, original issue referred to plug-in being loaded on each task run. 

Your issue is about plug-in being loaded twice (?) inside “webserver” component. 

I’m not sure how can this happen, do you have a way to reproduce?

**Comment by elephantum** [comment]
@twbecker please file a new issue with steps to reproduce pls
